src/providers/trusted_service/context.rs

```diff
@@ -0,0 +1,117 @@
+use std::io::{Error, ErrorKind};
+use std::ptr::null_mut;
+use ts_binding::*;
+
+#[allow(
+    non_snake_case,
+    non_camel_case_types,
+    non_upper_case_globals,
+    clippy::unseparated_literal_suffix,
+    // There is an issue where long double become u128 in extern blocks. Check this issue:
+    // https://github.com/rust-lang/rust-bindgen/issues/1549
+    improper_ctypes,
+    missing_debug_implementations,
+    trivial_casts,
+    clippy::all,
+    unused,
+    unused_qualifications
+)]
+pub mod ts_binding {
+    include!(concat!(env!("OUT_DIR"), "/ts_bindings.rs"));
+
+    unsafe impl Send for rpc_caller {}
+    unsafe impl Sync for rpc_caller {}
+
+    unsafe impl Send for service_context {}
+    unsafe impl Sync for service_context {}
+}
+
+#[allow(
+    non_snake_case,
+    non_camel_case_types,
+    non_upper_case_globals,
+    clippy::unseparated_literal_suffix,
+    // There is an issue where long double become u128 in extern blocks. Check this issue:
+    // https://github.com/rust-lang/rust-bindgen/issues/1549
+    improper_ctypes,
+    missing_debug_implementations,
+    trivial_casts,
+    clippy::all,
+    unused,
+    unused_qualifications
+)]
+mod ts_protobuf {
+    include!(concat!(env!("OUT_DIR"), "/ts_crypto.rs"));
+}
+
+#[derive(Debug)]
+pub struct Context {
+    rpc_caller: Option<rpc_caller>,
+    service_context: Option<service_context>,
+    rpc_session_handle: Option<MBox<c_void>>,
+}
+
+impl Context {
+    pub fn connect() -> anyhow::Result<Self> {
+        info!("Querying for crypto Trusted Services");
+        let mut status = 0;
+        let service_context = unsafe {
+            service_locator_query(
+                CString::new("sn:tf.org:crypto:0").unwrap().into_raw(),
+                &mut status,
+            )
+        };
+        if service_context == null_mut() || status != 0 {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!(
+                    "Failed to connect to Trusted Service; status code: {}",
+                    status
+                ),
+            )
+            .into());
+        }
+
+        info!("Starting crypto Trusted Service context");
+        let mut rpc_caller = null_mut();
+        let rpc_session_handle = unsafe { service_context_open(service_context, &mut rpc_caller) };
+        if rpc_caller == null_mut() || rpc_session_handle == null_mut() {
+            return Err(
+                Error::new(ErrorKind::Other, "Failed to start Trusted Service context").into(),
+            );
+        }
+        let rpc_caller_box = unsafe { MBox::from_raw(rpc_caller) };
+        let service_context_box = unsafe { MBox::from_raw(service_context) };
+        let rpc_session_handle = unsafe { MBox::from_raw(rpc_session_handle) };
+        let ctx = Context {
+            rpc_caller: Some(*rpc_caller_box),
+            service_context: Some(*service_context_box),
+            rpc_session_handle: Some(rpc_session_handle),
```

Comment:
When you dereference the `MBox` is that still keeping ownership? As in, is it not just copying the structure contents (which would be freed once this function ends)?

Reply:
Dereferencing copies the contents, I think, and keeps ownership of the initial object that came from the FFI. However, that object essentially contains some pointers. When the function ends, the `MBox` is dropped and, if I don't release them before, the pointers are dropped as well. But the FFI is the one who should be freeing them, ends up with a double free.

Reply:
ah I think I got it but why

> ends up with a double free.

?
If we copy all the contents of the `MBox` into owned structure then `MBox` will free the originals but it's fine because we made copies. For those copies, the FFI will free whatever needs to be freed and ownership will do the rest?

Reply:
Yeah, I think `MBox` is clever and tries to free the pointer we give it. Decided to just drop `MBox` usage completely and store the pointers instead. I'll implement `Send` and `Sync` for `Context` and we can have it lock every call with a mutex.